sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,240 @@
+
+So, overall the picture looks like:
+- device compiler generates property set/s which provide mapping
+  "unique symbolic ID" -> "various information required by DPC++ RT"
+- device or host compiler generates mapping
+  "address of a host variable" -> "unique symbolic ID" (as described below by
+  this document)
+- DPC++ RT uses these to mappings to obtain required information and somehow
+  uses it
+
+This design document describes two approaches of how the 
+"address of a host variable" -> "unique symbolic ID" mapping can be generated:
+the first one with integration footer and another one with modification of the
+host compiler.
+
+Both approaches have their pros and cons and they are expected to be implemented
+and exists in the implementation at the same time, but only one of them will be
+used at a time depending on whether 3rd-party host compiler is used or not.
+
+Integration footer can be used with 3rd-party host compilers, but it requires
+appending to a translation unit provided by user, which could affect debug
+information: since there are no compilers that support appending a file at the
+end (similar to `-include`), appending is done by generating a temporary input
+file using concatenation of the original input and integration footer.
+
+Such replacement of the main translation unit causes the following issues:
+- debug information about the source file might be incorrect, leading to
+  problems with gdb `l` command and code coverage tools
+- checksum of host and device source files becomes different which causes device
+  code debugging to be completely broken in some environments (such as MS Visual
+  Studio, for example)
+
+Customizing host compiler allows to avoid issues with debuggers and code
+coverage tools, but that is not an option if user wants to compile host part
+of an app with a 3rd-party host compiler.
+
+Further sections describe the implementation design of both approaches in more
+details, note that there are few components which should be modified regardless
+of which approach is in use.
+
+## Common front-end part
+
+DPC++ FE should support the following attribute:
+`[[__sycl_detail__::uniquely_identifiable_object(kind)]]`. This attribute accepts
+a string literal and should be applied to types (like `device_global` or
+`specialization_id`).
+
+Presence of the attribute instructs the compiler to perform the following
+things:
+- emit `sycl-unique-id` LLVM IR attribute on each definition of a variable of
+  type marked with `[[__sycl_detail__::uniquely_identifiable_object(kind)]]`
+  attribute. `sycl-unique-id` LLVM IR attribute should be accompanied by a
+  unique string identifier of a variable it is attached to. The rules for
+  creating this string are the same as for `__builtin_sycl_unique_stable_id` and
+  the same algorithm can be used when generating the string for the attribute
+- emit `sycl-uid-type` LLVM IR attribute alongside `sycl-unique-id`, which
```

Comment:
Is there a difference in later handling of spec constants and device globals?

Reply:
For the purposes of generating "host address" -> "unique symbolic ID" mapping no, there is no difference between spec constants and device global.

For the purposes of making those two features functional, yes, the further handling is different for those two, i.e. we have separate passes for spec constants and device globals and that won't change with this proposal.